src/qibo/backends/__init__.py

```diff
@@ -0,0 +1,113 @@
+from qibo import config
+from qibo.config import raise_error, log, warnings, BACKEND_NAME
+from qibo.backends.numpy import NumpyBackend
+from qibo.backends.tensorflow import TensorflowBackend
+
+
+CONSTRUCTED_BACKENDS = {}
+def construct_backend(name):
+    if name not in CONSTRUCTED_BACKENDS:
+        if name == "numpy":
+            CONSTRUCTED_BACKENDS["numpy"] = NumpyBackend()
+        elif name == "tensorflow":
+            CONSTRUCTED_BACKENDS["tensorflow"] = TensorflowBackend()
+        else:
+            raise_error(ValueError, "Unknown backend name {}.".format(name))
+    return CONSTRUCTED_BACKENDS.get(name)
+
+numpy_backend = construct_backend("numpy")
+numpy_matrices = numpy_backend.matrices
+
+if BACKEND_NAME == "tensorflow":
+    K = TensorflowBackend()
+else: # pragma: no cover
+    # CI uses tensorflow as default backend
+    K = NumpyBackend()
+
+def set_backend(backend="custom"):
+    """Sets backend used to implement gates.
+
+    Args:
+        backend (str): possible options are 'custom' for the gates that use
+            custom tensorflow operator and 'defaulteinsum' or 'matmuleinsum'
+            for the gates that use tensorflow primitives (``tf.einsum`` or
+            ``tf.matmul`` respectively).
+    """
+    if not config.ALLOW_SWITCHERS and backend != K.gates:
```

Comment:
What is the motivation for the `ALLOW_SWITCHERS`?

Reply:
It is a flag that "remembers" if a gate has been allocated so that if the user uses `set_backend` or `set_precision` after allocating gates, a warning is raised. It is just a warning so it doesn't have any particular effect on the code or results. 
Note that this exists in our current master and I think its behavior is not changed here. I am not sure if there is a better way to handle this warning.

Reply:
Good, thanks.